### PR TITLE
[MINOR][SQL] Move iterator.hasNext into try block in executeTask

### PR DESCRIPTION
**Pull request by viirya**
### What changes were proposed in this pull request?

This patch moves `iterator.hasNext` into the try block of `tryWithSafeFinallyAndFailureCallbacks` in `FileFormatWriter.executeTask`.

### Why are the changes needed?

Not only `dataWriter.writeWithIterator(iterator)` causes error, `iterator.hasNext` could cause error like:

```
org.apache.spark.shuffle.FetchFailedException: Block shuffle_1_106_21 is corrupted but checksum verification passed
```

As it is not wrapped in the try block, `abort` won't be called on the committer. But as `setupTask` is called, it is safer to call `abort` in any case that error happens after it.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing test

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala [hunk 1]
```diff
@@ -383,32 +383,38 @@ object FileFormatWriter extends Logging {
...
-              description, taskAttemptContext, committer, spec)
-          case _ =>
-            new DynamicPartitionDataSingleWriter(description, taskAttemptContext, committer)
-        }
-      }
+    var dataWriter: FileFormatDataWriter = null
 
     Utils.tryWithSafeFinallyAndFailureCallbacks(block = {
+      dataWriter =
+        if (sparkPartitionId != 0 && !iterator.hasNext) {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala:390]: When I looked at a customer issue, the following error happens

```
org.apache.spark.shuffle.FetchFailedException: Block shuffle_1_106_21 is corrupted but checksum verification passed
```

but I don't see the log of aborted task. It doesn't read a good taste.

Besides, as `setupTask` is called, it is safer to call `abort` in any error case.

**Comment by viirya** [comment]
cc @huaxingao @dongjoon-hyun

**Review by huaxingao — APPROVED** [review]
LGTM

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala [hunk 2]
```diff
@@ -383,32 +383,41 @@ object FileFormatWriter extends Logging {
...
       // If there is an error, abort the task
-      dataWriter.abort()
-      logError(log"Job ${MDC(JOB_ID, jobId)} aborted.")
+      if (dataWriter != null) {
+        dataWriter.abort()
+      } else {
+        committer.abortTask(taskAttemptContext)
+      }
+      logError(log"Job: ${MDC(JOB_ID, jobId)}, Task: ${MDC(TASK_ID, taskId)}, " +
+        log"Task attempt ${MDC(TASK_ATTEMPT_ID, taskAttemptId)} aborted.")
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/FileFormatWriter.scala:416]: We miss task and task attempt info previously. Adding them here.

**Review by huaxingao — APPROVED** [review]

**Comment by huaxingao** [comment]
Merged to master. Thanks @viirya

**Comment by viirya** [comment]
Thanks @huaxingao

**Pull request closed without merging**